coda-prototype/src/io/StorageUtils.js

```diff
@@ -0,0 +1,196 @@
+        });
+    }
+    clearActivityLog() {
+        return new Promise(function (resolve, reject) {
+            chrome.storage.local.remove("instrumentation", () => {
+                if (chrome.runtime.lastError) {
+                    console.log("Runtime error: Chrome failed reading from storage!");
+                    reject(chrome.runtime.lastError);
+                }
+                else {
+                    console.log("Cleared activity log!");
+                    resolve(true);
+                }
+            });
+        });
+    }
+    saveDataset(dataset) {
+        this.lastEdit = new Date();
+        chrome.storage.local.set({
+            "dataset": JSON.stringify(dataset),
+            "lastEdit": JSON.stringify(this.lastEdit)
+        }, () => {
+            console.log("Stored dataset edit timestamp: " + this.lastEdit);
+            chrome.storage.local.get((store) => {
+                let data = JSON.parse(store["dataset"]);
+                let datasetString = "dataset (schemes: "
+                    + Object.keys(data["schemes"]).length +
+                    ", events: " + Object.keys(data["events"]).length +
+                    ", sessions: " + Object.keys(data["sessions"]).length + ")";
+                console.log("In storage: Last edit (" + new Date(JSON.parse(store["lastEdit"])) + "), " + datasetString);
+                chrome.storage.local.getBytesInUse((bytesUnUse) => {
+                    console.log("Bytes in use: " + bytesUnUse);
+                    console.log("QUOTA_BYTES: " + chrome.storage.local.QUOTA_BYTES);
+                });
+            });
+        });
+    }
+    saveActivity(logEvent, uid) {
+        // save user activity in storage for instrumentation
+        if (logEvent.category.length != 0 && (logEvent.message.length > 0 || logEvent.data.length > 0) && logEvent.timestamp instanceof Date) {
+            activity.push(logEvent);
+            console.log("INSTRUMENTATION: " + logEvent.category + ":" + logEvent.message + ", stack size: " + activity.length);
+            if (activity.length % StorageManager._MAX_ACTIVITY_SAVE_FREQ == 0) {
+                chrome.storage.local.get("instrumentation", data => {
+                    if (chrome.runtime.lastError) {
+                        console.log(chrome.runtime.lastError);
+                    }
+                    else {
+                        let instr;
+                        if (data["instrumentation"]) {
+                            instr = JSON.parse(data["instrumentation"]).concat(activity);
+                        }
+                        else {
+                            instr = activity;
+                        }
+                        chrome.storage.local.set({ "instrumentation": JSON.stringify(instr), }, () => {
+                            if (chrome.runtime.lastError) {
+                                console.log(chrome.runtime.lastError);
+                            }
+                            else {
```

Comment:
I find the deep nested if-else blocks very hard to read. Can they be flattened into an easier pattern?